utils/log_parser.py

```diff
@@ -15,136 +15,214 @@
 
 import argparse
 import re
+from typing import (
+    List,
+    Dict,
+    Tuple,
+)
 
 # Maintaining all nodes
-NODES_MAP = {}
-
-# Maintaining all edges
-EDGES_MAP = {}
+NODES_MAP: Dict[Tuple[str, str], 'Node'] = {}
 
 # Scope stack
-SCOPE_STACK = []
+SCOPE_STACK: List[str] = []
 
 
 class Node:
-    def __init__(self, kindName, name):
-        self.kindName_ = kindName
-        self.name_ = name
-        self.inputs_ = []
-        self.users_ = {}
+    """A class that represents a node in the compute graph
+
+    Public attributes:
+        kindName_: str. The kind name.
+        name_: str. The node name.
+        inputs_: List[Tuple['Node', int]]. Input node values.
+        users_: Dict['Node', int]. The users of this node.
+    """
+
+    def __init__(self, kindName: str, name: str):
+        self.kindName_: str = kindName
+        self.name_: str = name
+        self.inputs_: List[Tuple['Node', int]] = []
```

Comment:
Consider creating `NamedTuple` for this as well, e.g.:
```
class NodeValue(typing.NamedTuple):
  node: 'Node'
  resNo: int
```

Then, you can annotate the type of `inputs_` as `List[NodeValue]`.